### PR TITLE
`Integrated code lifecycle`: Show difference between commits

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently we don't show a commit history but rather a submission history. This means if one submission consists of many commits, only the last commit would show all the changes. Instead, we want to show the changes for each commit separately. 

### Description

On the server side, an extra `CommitHistoryService` was added. The parser for that is used for git diff reports was refactored to it's own class so that  `CommitHistoryService` and `ProgrammingExerciseGitDiffReportService` can use it. 

On the client side, mainly service methods had to be refactored, and the `commit-details-view.component`. Mainly, the submission logic was removed, and only the commits are handled now.

### Steps for Testing

Prerequisites:

**🚨🚨🚨Only deploy on ts3, ts6 and localci-staging🚨🚨🚨**

- 1 Instructor
- 1 Students
- 1 Programming Exercise 

1. Log in to Artemis
2. Navigate to a Programming Exercise
3. Clone the repository and open it in your local IDE 
4. Commit a few times, then push your code
5. In the Programming Exercise, navigate to the Repository > Commit History
6. Check that all commits are shown (only the most recent one has a result)
7. Navigate to the diff of every commit by clicking on the commit names.
8. Check that the diffs are according to your changes for each commit
9. Also check that you can see the diff for the template commit (first commit in commit history)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [git-diff-report.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts.html) | 100% | ✅ |
| [programming-exercise.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/services/programming-exercise.service.ts.html) | 90.25% | ✅ ❌ |
| [commit-details-view.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Client-Tests/app/localvc/commit-details-view/commit-details-view.component.ts.html) | 100% | ✅ |
| [repository-view.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Client-Tests/app/localvc/repository-view/repository-view.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExerciseGitDiffReportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.hestia/ProgrammingExerciseGitDiffReportService.html) | 81% | ✅ ❌ |
| [GitDiffReportParser.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/GitDiffReportParser.html) | 83% | ✅ ❌ |
| [ProgrammingExerciseGitDiffReportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.hestia/ProgrammingExerciseGitDiffReportResource.html) | 95% | ✅ ❌ |
| [CommitHistoryService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6132/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.localvc/CommitHistoryService.html) | 90% | ✅ ❌ |

### Screenshots

Screenshot 2024-03-06 at 18 59 14

Screenshot 2024-03-06 at 18 59 24

Screenshot 2024-03-06 at 18 59 43

## Summary by CodeRabbit

- **New Features**
    - Introduced a `GitDiffReportParser` for enhanced parsing of git-diff outputs, improving the accuracy of diff reports.
    - Added functionality to generate git-diff reports for programming exercise participations between two commit hashes or for a single commit.
    - Updated user interface components to better display git-diff information and handle different repository views.
- **Enhancements**
    - Refined the logic for displaying messages in the git-diff report component based on commit and repository conditions.
    - Enhanced the `ProgrammingExerciseService` to retrieve git-diff reports based on specific commits, improving the relevance of the information displayed.
- **Bug Fixes**
    - Corrected display and logic issues in the commit details and repository view components.
- **Tests**
    - Added and updated integration and component tests to cover new git-diff functionality and commit history retrieval.
- **Refactor**
    - Significant refactoring in handling git diffs, including the introduction of new services and components for better separation of concerns and improved maintainability.
- **Chores**
    - Removed obsolete methods and properties related to the old submission-based diff report retrieval, streamlining the codebase.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 1]
```diff
@@ -273,6 +273,29 @@ public ProgrammingExerciseGitDiffReport generateReportForSubmissions(Programming
...
+        var repositoryUri = participation.getVcsRepositoryUri();
+        var repo1 = gitService.getOrCheckoutRepository(repositoryUri, true);
+        var repo1Path = repo1.getLocalPath();
+        var repo2Path = fileService.getTemporaryUniqueSubfolderPath(repo1Path.getParent(), 5);
+        FileSystemUtils.copyRecursively(repo1Path, repo2Path);
+        repo1 = gitService.checkoutRepositoryAtCommit(repo1, commitHash1);
+        var repo2 = gitService.getExistingCheckedOutRepositoryByLocalPath(repo2Path, repositoryUri);
+        repo2 = gitService.checkoutRepositoryAtCommit(repo2, commitHash2);
+        return parseFilesAndCreateReport(repo1, repo2);
+    }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:297]: This is duplicated from above. You can change the method `generateReportForSubmissions` to:
```java
return generateReportForCommits((ProgrammingExerciseParticipation) submission1.getParticipation(), submission1.getCommitHash(), submission2.getCommitHash());
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/hestia/ProgrammingExerciseGitDiffReportResource.java [hunk 2]
```diff
@@ -128,61 +131,26 @@ public ResponseEntity<ProgrammingExerciseGitDiffReportDTO> getGitDiffReportForSu
...
-                    "exerciseIdsMismatch");
-        }
-        participationAuthCheckService.checkCanAccessParticipationElseThrow(submission.getParticipation());
-        var report = gitDiffReportService.createReportForSubmissionWithTemplate(exercise, submission);
+    public ResponseEntity<ProgrammingExerciseGitDiffReportDTO> getGitDiffReportForCommits(@PathVariable long participationId, @PathVariable String commitHash1,
+            @PathVariable String commitHash2) throws GitAPIException, IOException {
+        log.debug("REST request to get a ProgrammingExerciseGitDiffReport for two commits for commit {} and commit {} of participation {}", commitHash1, commitHash2,
+                participationId);
+
+        ProgrammingExerciseStudentParticipation participation = programmingExerciseStudentParticipationRepository.findByIdElseThrow(participationId);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/hestia/ProgrammingExerciseGitDiffReportResource.java:151]: Missing consistency-check between the exerciseId and the exercise of the participation.

File: src/main/java/de/tum/in/www1/artemis/web/rest/GitDiffReportParser.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                    case ' ' -> handleUnchanged(parserState);
+                    default -> parserState.deactivateCodeReading = true;
+                }
+            }
+        }
+        if (!parserState.currentEntry.isEmpty()) {
+            parserState.entries.add(parserState.currentEntry);
+        }
+        return parserState.entries;
+    }
```
[the remark this answers is not in the document]
Comment (mateusmm01) [reply src/main/java/de/tum/in/www1/artemis/web/rest/GitDiffReportParser.java:50]: This isn't my code I just moved it

**Review by laurenzfb — APPROVED** [review]
tested in testing session

**Review by reschandreas — APPROVED** [review]
demonstrated in testing session, works as expected

**Review by milljoniaer — APPROVED** [review]
Code LGTM

**Review by rstief — APPROVED** [review]
tested on ts3 works as expected

**Pull request merged**